lmp/model/_transformer.py

```diff
@@ -0,0 +1,199 @@
+import math
+import torch
+import numpy as np
+from torch.utils.tensorboard import SummaryWriter
+from torch.nn import Linear
+from torch.nn.functional import softmax, relu
+
+
+class PositionalEncoding(torch.nn.Module):
+    "Implement the PE function."
+
+    def __init__(self, d_model, dropout, max_len=5000):
+        super(PositionalEncoding, self).__init__()
+        self.dropout = torch.nn.Dropout(p=dropout)
+
+        # Compute the positional encodings once in log space.
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2) *
+                             -(math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        x = x + torch.autograd.Variable(self.pe[:, :x.size(-2)],
+                                        requires_grad=False)
+        return self.dropout(x)
+
+
+class Attention(torch.nn.Module):
+    """
+    Attention
+    (B, L, E) (B, L, E) (B, L, E) -> (B, L, D)
+    """
+
+    def __init__(self, E, D):
+        super().__init__()
+        self.Wq = Linear(E, D)
+        self.Wk = Linear(E, D)
+        self.Wv = Linear(E, D)
+        self.D = D
+
+    def forward(self, Q, K, V, mask):
+        Q = self.Wq(Q)
+        K = self.Wk(K)
+        V = self.Wv(V)
+        Z = softmax(
+            (Q @ K.transpose(-2, -1) / math.sqrt(self.D)).masked_fill(mask == 0, -1e9), dim=-1) @ V
+        return Z
+
+
+class MultiHeadAttention(torch.nn.Module):
+    """
+    Multi Head Attention:
+
+    concatenate multiple attention and reduce results by linear transform
```

Comment:
Write every comment as a complete sentence. (You need to capitalize every sentence and put colon.